### PR TITLE
FEATURE: Category reordering dialog

### PR DESCRIPTION
**Pull request by riking**
- only shown if fixed_category_positions is turned on

### Button
image
Mobile
image

## Dialog
image
image

### CODE REVIEW

File: app/assets/javascripts/discourse/controllers/reorder-categories.js.es6 [hunk 1]
```diff
@@ -0,0 +1,106 @@
+import ModalFunctionality from 'discourse/mixins/modal-functionality';
+import { bufferedProperty } from 'discourse/mixins/buffered-content';
+import binarySearch from 'discourse/lib/binary-search';
+import { popupAjaxError } from 'discourse/lib/ajax-error';
+
+// Modal for editing / creating a category
+export default Ember.Controller.extend(ModalFunctionality, Ember.Evented, {
+
+  rawCategories: function() {
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/controllers/reorder-categories.js.es6]: I believe you can use an alias here: `rawCategories: Ember.computed.alias("site.categories")`

File: app/assets/javascripts/discourse/initializers/message-bus.js.es6 [hunk 2]
```diff
@@ -20,5 +22,25 @@ export default {
     messageBus.alwaysLongPoll = Discourse.Environment === "development";
     messageBus.start();
     Discourse.KeyValueStore.init("discourse_", messageBus);
+
+    messageBus.callbackInterval = siteSettings.anon_polling_interval;
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/initializers/message-bus.js.es6]: Not sure this is related to this PR. Can you do that in another commit?

File: app/assets/javascripts/discourse/templates/modal/reorder-categories.hbs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      <th class="th-pos">Position</th>
+      <th class="th-cat">Category</th>
+      </thead>
+      {{#each categoriesGrouped as |group|}}
+        <tbody>
+        {{#each group.cats as |cat|}}
+          <tr data-category-id="{{cat.id}}">
+            <td>
+              {{number-field number=cat.position}}
+              <button class="btn no-text" {{action "moveUp" cat}}>{{fa-icon "arrow-up"}}</button>
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/templates/modal/reorder-categories.hbs]: You should use our `{{d-button}}` component instead ;)

Comment (riking) [inline app/assets/javascripts/discourse/templates/modal/reorder-categories.hbs]: I need to pass a parameter to the `action`.

File: app/assets/javascripts/discourse/components/number-field.js.es6 [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import computed from 'ember-addons/ember-computed-decorators';
+
+export default Ember.TextField.extend({
+
+  @computed('number')
+  value: {
+    get() {
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/components/number-field.js.es6]: you could do `get(number)` here :smile:

File: app/assets/javascripts/discourse/components/number-field.js.es6 [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+      return parseInt(number);
+    },
+    set(value) {
+      const num = parseInt(value);
+      this.set('number', num);
+      return num.toString();
+    }
+  },
+
+  placeholder: function() {
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/components/number-field.js.es6]: This could be:

```javascript
@computed("placeholderKey")
placeholder(key) {
  return key ? I18n.t(key) : "":
}
```

File: app/assets/javascripts/discourse/controllers/reorder-categories.js.es6 [hunk 6]
```diff
@@ -0,0 +1,102 @@
+import ModalFunctionality from 'discourse/mixins/modal-functionality';
+import { bufferedProperty } from 'discourse/mixins/buffered-content';
+import binarySearch from 'discourse/lib/binary-search';
+import { popupAjaxError } from 'discourse/lib/ajax-error';
+
+// Modal for editing / creating a category
+export default Ember.Controller.extend(ModalFunctionality, Ember.Evented, {
+
+  categoriesBuffered: function() {
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/controllers/reorder-categories.js.es6]: Again, ES6 FTW ;)

```javascript
@computed("site.categories")
categoriesBuffered(categories) {
   // ...
}
```

File: app/assets/javascripts/discourse/controllers/reorder-categories.js.es6 [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    const bufProxy = Em.ObjectProxy.extend(BufferedProxy);
+    return this.get('site.categories').map(function(c) {
+      return bufProxy.create({ content: c });
+    });
+  }.property('site.categories'),
+
+  // uses propertyDidChange()
+  categoriesGrouped: function(key, newValue) {
+    if (arguments.length > 1) {
+      debugger;
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/controllers/reorder-categories.js.es6]: Ooopps ;)

File: app/assets/javascripts/discourse/controllers/reorder-categories.js.es6 [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    },
+    moveDown: function(cat) {
+      this.moveDir(cat, 1);
+    },
+
+    commit: function() {
+      this.get('categoriesBuffered').forEach(function(bc) {
+        if (bc.get('hasBufferedChanges')) {
+          bc.applyBufferedChanges();
+          //bc.set('hasBufferedChanges', false);
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/controllers/reorder-categories.js.es6]: Is that usefull?

File: app/assets/javascripts/discourse/controllers/reorder-categories.js.es6 [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+      this.propertyDidChange('categoriesGrouped');
+    },
+
+    saveOrder: function() {
+      const data = {},
+        self = this;
+      this.get('categoriesBuffered').forEach(function(cat) {
+        data[cat.get('id')] = cat.get('position');
+      });
+      Discourse.ajax('/categories/reorder', {type: 'POST', data: {mapping: JSON.stringify(data)}}).then(function() {
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/controllers/reorder-categories.js.es6]:
```javascript
Discourse.ajax('/categories/reorder', {type: 'POST', data: {mapping: JSON.stringify(data)}}).then(() => this.send("closeModal")).catch(popupAjaxError);
```

File: app/assets/javascripts/discourse/templates/modal/reorder-categories.hbs (hunk 3 above)
Comment (ZogStriP) [inline app/assets/javascripts/discourse/templates/modal/reorder-categories.hbs]: You can use "`actionParam`" ;)

File: app/assets/javascripts/discourse/controllers/reorder-categories.js.es6 (hunk 8 above)
Comment (riking) [inline app/assets/javascripts/discourse/controllers/reorder-categories.js.es6]: Nope, was leftover from when the `if` statement was checking the wrong property...

File: app/assets/javascripts/discourse/controllers/reorder-categories.js.es6 [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+  @computed("site.categories")
+  categoriesBuffered(categories) {
+    const bufProxy = Em.ObjectProxy.extend(BufferedProxy);
+    return categories.map(function(c) {
+      return bufProxy.create({ content: c });
+    });
+  },
+
+  // uses propertyDidChange()
+  categoriesGrouped: function(key, newValue) {
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/controllers/reorder-categories.js.es6]: ES6 => `categoriesGrouped(key, newValue) {`

File: app/assets/stylesheets/common/admin/cat_reorder.scss [hunk 11]
```diff
@@ -0,0 +1,28 @@
+.reorder-categories {
+  input {
+    width: 4em;
+  }
+  .th-pos {
+    width: calc(4em + 150px);
+  }
+  tbody tr {
+    background-color: transparent;
+    transition: background 0s cubic-bezier(0.55, 0.06, 0.68, 0.19);
```
Comment (SamSaffron) [inline app/assets/stylesheets/common/admin/cat_reorder.scss]: you got to use the transition helper or it will not work in iphone/ipad

**Comment by SamSaffron** [comment]
hold off on PR while we figure out why fixed_category_positions was removed

**Comment by SamSaffron** [comment]
yay :confetti_ball:
